dandi-api tests against dandi-cli started to fail in staging

- master seems to be fine: https://github.com/dandi/dandi-api/actions/workflows/cli-integration.yml?query=branch%3Amaster . 
- [ ] submitted https://github.com/dandi/dandi-api/pull/366  to simplify archaeological expeditions in the future
- for a proper solution IMHO we need dandi-api to
  - [ ] show a proper version - have https://github.com/dandi/dandi-api/issues/222 addressed
  - [ ] boost version whenever a breaking change is introduced https://github.com/dandi/dandi-api/issues/368
  - [ ] dandi-cli to code in different logic depending on API version
- but may be there is a simple workaround for now (didn't look in detail of failures) ?
- might relate to #677 although there it worked fine for me

<details>
<summary>cut/paste from one of the broken runs</summary> 

```shell
2021-06-23T23:40:13.7802778Z _________________________ test_download_newest_version _________________________
2021-06-23T23:40:13.7803114Z 
2021-06-23T23:40:13.7804913Z local_dandi_api = {'api_key': '4e15b41f88e021a134c2a8bef3fdb0975794e56c', 'client': <dandi.dandiapi.DandiAPIClient object at 0x7f3d31374...di_instance(gui=None, redirector=None, api='http://localhost:8000/api'), 'instance_id': 'dandi-api-local-docker-tests'}
2021-06-23T23:40:13.7807532Z text_dandiset = {'client': <dandi.dandiapi.DandiAPIClient object at 0x7f3d31374970>, 'dandiset_id': '000028', 'dspath': PosixPath('/tm...-of-runner/pytest-0/text_dandiset23'), 'reupload': <function text_dandiset.<locals>.upload_dandiset at 0x7f3d300f7280>}
2021-06-23T23:40:13.7809214Z tmp_path = PosixPath('/tmp/pytest-of-runner/pytest-0/test_download_newest_version0')
2021-06-23T23:40:13.7809742Z 
2021-06-23T23:40:13.7810278Z     def test_download_newest_version(local_dandi_api, text_dandiset, tmp_path):
2021-06-23T23:40:13.7810917Z         dandiset_id = text_dandiset["dandiset_id"]
2021-06-23T23:40:13.7811753Z         download(f"{local_dandi_api['instance'].api}/dandisets/{dandiset_id}", tmp_path)
2021-06-23T23:40:13.7812495Z         assert (tmp_path / dandiset_id / "file.txt").read_text() == "This is test text.\n"
2021-06-23T23:40:13.7813190Z >       text_dandiset["client"].publish_version(dandiset_id, "draft")
2021-06-23T23:40:13.7813583Z 
2021-06-23T23:40:13.7814400Z /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/tests/test_download.py:134: 
2021-06-23T23:40:13.7815095Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-06-23T23:40:13.7815989Z /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/dandiapi.py:628: in publish_version
2021-06-23T23:40:13.7816714Z     return self.post(
2021-06-23T23:40:13.7817657Z /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/dandiapi.py:213: in post
2021-06-23T23:40:13.7818364Z     return self.send_request(
2021-06-23T23:40:13.7818794Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-06-23T23:40:13.7819045Z 
2021-06-23T23:40:13.7819852Z self = <dandi.dandiapi.DandiAPIClient object at 0x7f3d31374970>, method = 'POST'
2021-06-23T23:40:13.7820887Z path = '/dandisets/000028/versions/draft/publish/', parameters = {}, data = None
2021-06-23T23:40:13.7821575Z files = None, json = None, headers = None, json_resp = True, retry = None
2021-06-23T23:40:13.7822068Z kwargs = {}
2021-06-23T23:40:13.7822732Z f = <bound method Session.post of <requests.sessions.Session object at 0x7f3d3148c730>>
2021-06-23T23:40:13.7823807Z url = 'http://localhost:8000/api/dandisets/000028/versions/draft/publish/'
2021-06-23T23:40:13.7825175Z _headers = {'Authorization': 'token 4e15b41f88e021a134c2a8bef3fdb0975794e56c', 'User-Agent': 'dandi/0.20.0 requests/2.25.1 CPython/3.8.10', 'accept': 'application/json'}
2021-06-23T23:40:13.7826283Z doretry = <tenacity.retry.retry_any object at 0x7f3d2bfd5c10>
2021-06-23T23:40:13.7826874Z result = <Response [400]>
2021-06-23T23:40:13.7827784Z msg = 'Error 400 while sending POST request to http://localhost:8000/api/dandisets/000028/versions/draft/publish/'
2021-06-23T23:40:13.7828352Z 
2021-06-23T23:40:13.7828693Z     def send_request(
2021-06-23T23:40:13.7829042Z         self,
2021-06-23T23:40:13.7829367Z         method,
2021-06-23T23:40:13.7829684Z         path,
2021-06-23T23:40:13.7830052Z         parameters=None,
2021-06-23T23:40:13.7830419Z         data=None,
2021-06-23T23:40:13.7830858Z         files=None,
2021-06-23T23:40:13.7831189Z         json=None,
2021-06-23T23:40:13.7831548Z         headers=None,
2021-06-23T23:40:13.7831908Z         json_resp=True,
2021-06-23T23:40:13.7832267Z         retry=None,
2021-06-23T23:40:13.7832596Z         **kwargs,
2021-06-23T23:40:13.7832905Z     ):
2021-06-23T23:40:13.7833177Z         """
2021-06-23T23:40:13.7833721Z         This method looks up the appropriate method, constructs a request URL
2021-06-23T23:40:13.7834481Z         from the base URL, path, and parameters, and then sends the request. If
2021-06-23T23:40:13.7835178Z         the method is unknown or if the path is not found, an exception is
2021-06-23T23:40:13.7835875Z         raised, otherwise a JSON object is returned with the response.
2021-06-23T23:40:13.7836357Z     
2021-06-23T23:40:13.7836862Z         This is a convenience method to use when making basic requests that do
2021-06-23T23:40:13.7837594Z         not involve multipart file data that might need to be specially encoded
2021-06-23T23:40:13.7838204Z         or handled differently.
2021-06-23T23:40:13.7838564Z     
2021-06-23T23:40:13.7839059Z         :param method: The HTTP method to use in the request (GET, POST, etc.)
2021-06-23T23:40:13.7839603Z         :type method: str
2021-06-23T23:40:13.7840149Z         :param path: A string containing the path elements for this request.
2021-06-23T23:40:13.7841237Z             Note that the path string should not begin or end with the path  separator, '/'.
2021-06-23T23:40:13.7841793Z         :type path: str
2021-06-23T23:40:13.7842368Z         :param parameters: A dictionary mapping strings to strings, to be used
2021-06-23T23:40:13.7843046Z             as the key/value pairs in the request parameters.
2021-06-23T23:40:13.7843576Z         :type parameters: dict
2021-06-23T23:40:13.7844352Z         :param data: A dictionary, bytes or file-like object to send in the body.
2021-06-23T23:40:13.7845357Z         :param files: A dictionary of 'name' => file-like-objects for multipart encoding upload.
2021-06-23T23:40:13.7846017Z         :type files: dict
2021-06-23T23:40:13.7846499Z         :param json: A JSON object to send in the request body.
2021-06-23T23:40:13.7846979Z         :type json: dict
2021-06-23T23:40:13.7847557Z         :param headers: If present, a dictionary of headers to encode in the request.
2021-06-23T23:40:13.7848189Z         :type headers: dict
2021-06-23T23:40:13.7848860Z         :param json_resp: Whether the response should be parsed as JSON. If False, the raw
2021-06-23T23:40:13.7849658Z             response object is returned. To get the raw binary content of the response,
2021-06-23T23:40:13.7850375Z             use the ``content`` attribute of the return value, e.g.
2021-06-23T23:40:13.7850814Z     
2021-06-23T23:40:13.7851541Z             .. code-block:: python
2021-06-23T23:40:13.7851886Z     
2021-06-23T23:40:13.7852499Z                 resp = client.get('my/endpoint', json_resp=False)
2021-06-23T23:40:13.7853095Z                 print(resp.content)  # Raw binary content
2021-06-23T23:40:13.7853676Z                 print(resp.headers)  # Dict of headers
2021-06-23T23:40:13.7854082Z     
2021-06-23T23:40:13.7854429Z         :type json_resp: bool
2021-06-23T23:40:13.7855001Z         :param retry: an optional tenacity `retry` argument for retrying the
2021-06-23T23:40:13.7855540Z             request method
2021-06-23T23:40:13.7855883Z         """
2021-06-23T23:40:13.7856229Z         if not parameters:
2021-06-23T23:40:13.7856641Z             parameters = {}
2021-06-23T23:40:13.7856969Z     
2021-06-23T23:40:13.7857347Z         # Look up the HTTP method we need
2021-06-23T23:40:13.7857810Z         f = self._request_func(method)
2021-06-23T23:40:13.7858185Z     
2021-06-23T23:40:13.7858526Z         url = self.get_url(path)
2021-06-23T23:40:13.7858877Z     
2021-06-23T23:40:13.7859386Z         # Make the request, passing parameters and authentication info
2021-06-23T23:40:13.7860004Z         _headers = dict(self._headers)
2021-06-23T23:40:13.7860417Z         if headers:
2021-06-23T23:40:13.7860928Z             _headers.update(headers)
2021-06-23T23:40:13.7861325Z     
2021-06-23T23:40:13.7861711Z         if json_resp and "accept" not in _headers:
2021-06-23T23:40:13.7862251Z             _headers["accept"] = "application/json"
2021-06-23T23:40:13.7862654Z     
2021-06-23T23:40:13.7863064Z         lgr.debug("%s %s", method.upper(), url)
2021-06-23T23:40:13.7863465Z     
2021-06-23T23:40:13.7864177Z         # urllib3's ConnectionPool isn't thread-safe, so we sometimes hit
2021-06-23T23:40:13.7864948Z         # ConnectionErrors on the start of an upload.  Retry when this happens.
2021-06-23T23:40:13.7865704Z         # Cf. <https://github.com/urllib3/urllib3/issues/951>.
2021-06-23T23:40:13.7866371Z         doretry = tenacity.retry_if_exception_type(
2021-06-23T23:40:13.7866991Z             requests.ConnectionError
2021-06-23T23:40:13.7867635Z         ) | tenacity.retry_if_result(lambda r: r.status_code == 503)
2021-06-23T23:40:13.7868144Z         if retry is not None:
2021-06-23T23:40:13.7868547Z             doretry |= retry
2021-06-23T23:40:13.7868865Z     
2021-06-23T23:40:13.7869155Z         try:
2021-06-23T23:40:13.7869552Z             result = try_multiple(5, doretry, 1.1)(
2021-06-23T23:40:13.7870238Z                 f,
2021-06-23T23:40:13.7870543Z                 url,
2021-06-23T23:40:13.7870938Z                 params=parameters,
2021-06-23T23:40:13.7871349Z                 data=data,
2021-06-23T23:40:13.7871704Z                 files=files,
2021-06-23T23:40:13.7872068Z                 json=json,
2021-06-23T23:40:13.7872442Z                 headers=_headers,
2021-06-23T23:40:13.7872825Z                 **kwargs,
2021-06-23T23:40:13.7873133Z             )
2021-06-23T23:40:13.7873490Z         except Exception:
2021-06-23T23:40:13.7873992Z             lgr.exception("HTTP connection failed")
2021-06-23T23:40:13.7874454Z             raise
2021-06-23T23:40:13.7874743Z     
2021-06-23T23:40:13.7875187Z         lgr.debug("Response: %d", result.status_code)
2021-06-23T23:40:13.7875619Z     
2021-06-23T23:40:13.7876172Z         # If success, return the json object. Otherwise throw an exception.
2021-06-23T23:40:13.7876824Z         if not result.ok:
2021-06-23T23:40:13.7877498Z             msg = f"Error {result.status_code} while sending {method} request to {url}"
2021-06-23T23:40:13.7878159Z             if result.status_code == 409:
2021-06-23T23:40:13.7898212Z                 # Blob exists on server; log at DEBUG level
2021-06-23T23:40:13.7898823Z                 lgr.debug("%s: %s", msg, result.text)
2021-06-23T23:40:13.7899253Z             else:
2021-06-23T23:40:13.7899702Z                 lgr.error("%s: %s", msg, result.text)
2021-06-23T23:40:13.7900308Z >           raise requests.HTTPError(msg, response=result)
2021-06-23T23:40:13.7901432Z E           requests.exceptions.HTTPError: Error 400 while sending POST request to http://localhost:8000/api/dandisets/000028/versions/draft/publish/
2021-06-23T23:40:13.7902244Z 
2021-06-23T23:40:13.7903233Z /opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/dandiapi.py:177: HTTPError
2021-06-23T23:40:13.7904280Z ------------------------------ Captured log setup ------------------------------
2021-06-23T23:40:13.7905062Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/dandisets/
2021-06-23T23:40:13.7905745Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7906422Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/auth/token
2021-06-23T23:40:13.7907084Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7907922Z DEBUG    dandi:dandiset.py:99 Found identifier 000028 in top level 'identifier'
2021-06-23T23:40:13.7908589Z INFO     dandi:upload.py:75 Found 5 files to consider
2021-06-23T23:40:13.7909371Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7910127Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7911149Z ERROR    dandi:upload.py:249 Failed to extract metadata from /tmp/pytest-of-runner/pytest-0/text_dandiset23/file.txt
2021-06-23T23:40:13.7912044Z Traceback (most recent call last):
2021-06-23T23:40:13.7913143Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/upload.py", line 245, in process_path
2021-06-23T23:40:13.7914301Z     asset_metadata = nwb2asset(
2021-06-23T23:40:13.7915350Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 670, in nwb2asset
2021-06-23T23:40:13.7916137Z     metadata = get_metadata(nwb_path)
2021-06-23T23:40:13.7917201Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 116, in fingerprinter
2021-06-23T23:40:13.7918058Z     ret = fingerprinted(path, *args, **kwargs_)
2021-06-23T23:40:13.7919084Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 591, in __call__
2021-06-23T23:40:13.7919865Z     return self._cached_call(args, kwargs)[0]
2021-06-23T23:40:13.7921707Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 534, in _cached_call
2021-06-23T23:40:13.7922522Z     out, metadata = self.call(*args, **kwargs)
2021-06-23T23:40:13.7923552Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 761, in call
2021-06-23T23:40:13.7924286Z     output = self.func(*args, **kwargs)
2021-06-23T23:40:13.7925349Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 84, in fingerprinted
2021-06-23T23:40:13.7926155Z     return f(path, *args, **kwargs)
2021-06-23T23:40:13.7927153Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 51, in get_metadata
2021-06-23T23:40:13.7927957Z     meta["nwb_version"] = get_nwb_version(path)
2021-06-23T23:40:13.7929009Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/pynwb_utils.py", line 95, in get_nwb_version
2021-06-23T23:40:13.7929812Z     with h5py.File(filepath, "r") as h5file:
2021-06-23T23:40:13.7930828Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 406, in __init__
2021-06-23T23:40:13.7931592Z     fid = make_fid(name, mode, userblock_size,
2021-06-23T23:40:13.7932717Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 173, in make_fid
2021-06-23T23:40:13.7933481Z     fid = h5f.open(name, flags, fapl=fapl)
2021-06-23T23:40:13.7934148Z   File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.7934957Z   File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.7935639Z   File "h5py/h5f.pyx", line 88, in h5py.h5f.open
2021-06-23T23:40:13.7936205Z OSError: Unable to open file (file signature not found)
2021-06-23T23:40:13.7937528Z DEBUG    dandi:dandiapi.py:454 Calculated dandi-etag of d5b215bbddf463221d1cc55c31098d40-1 for /tmp/pytest-of-runner/pytest-0/text_dandiset23/file.txt
2021-06-23T23:40:13.7938598Z DEBUG    dandi:dandiapi.py:463 file.txt: Beginning upload
2021-06-23T23:40:13.7939386Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/uploads/initialize/
2021-06-23T23:40:13.7940110Z DEBUG    dandi:dandiapi.py:167 Response: 409
2021-06-23T23:40:13.7940995Z DEBUG    dandi:dandiapi.py:174 Error 409 while sending POST request to http://localhost:8000/api/uploads/initialize/: "Blob already exists."
2021-06-23T23:40:13.7941965Z DEBUG    dandi:dandiapi.py:478 file.txt: Blob already exists on server
2021-06-23T23:40:13.7942714Z DEBUG    dandi:dandiapi.py:551 file.txt: Assigning asset blob to dandiset & version
2021-06-23T23:40:13.7943597Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7944333Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7945098Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7945870Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7946576Z INFO     dandi:dandiapi.py:565 file.txt: Asset successfully uploaded
2021-06-23T23:40:13.7947428Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7948203Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7949336Z ERROR    dandi:upload.py:249 Failed to extract metadata from /tmp/pytest-of-runner/pytest-0/text_dandiset23/subdir2/coconut.txt
2021-06-23T23:40:13.7950169Z Traceback (most recent call last):
2021-06-23T23:40:13.7951199Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/upload.py", line 245, in process_path
2021-06-23T23:40:13.7951979Z     asset_metadata = nwb2asset(
2021-06-23T23:40:13.7952975Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 670, in nwb2asset
2021-06-23T23:40:13.7953755Z     metadata = get_metadata(nwb_path)
2021-06-23T23:40:13.7954811Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 116, in fingerprinter
2021-06-23T23:40:13.7955653Z     ret = fingerprinted(path, *args, **kwargs_)
2021-06-23T23:40:13.7956682Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 591, in __call__
2021-06-23T23:40:13.7957445Z     return self._cached_call(args, kwargs)[0]
2021-06-23T23:40:13.7958471Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 534, in _cached_call
2021-06-23T23:40:13.7959269Z     out, metadata = self.call(*args, **kwargs)
2021-06-23T23:40:13.7960250Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 761, in call
2021-06-23T23:40:13.7961111Z     output = self.func(*args, **kwargs)
2021-06-23T23:40:13.7962167Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 84, in fingerprinted
2021-06-23T23:40:13.7962949Z     return f(path, *args, **kwargs)
2021-06-23T23:40:13.7963956Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 51, in get_metadata
2021-06-23T23:40:13.7964753Z     meta["nwb_version"] = get_nwb_version(path)
2021-06-23T23:40:13.7965865Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/pynwb_utils.py", line 95, in get_nwb_version
2021-06-23T23:40:13.7966689Z     with h5py.File(filepath, "r") as h5file:
2021-06-23T23:40:13.7967692Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 406, in __init__
2021-06-23T23:40:13.7968465Z     fid = make_fid(name, mode, userblock_size,
2021-06-23T23:40:13.7969482Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 173, in make_fid
2021-06-23T23:40:13.7970230Z     fid = h5f.open(name, flags, fapl=fapl)
2021-06-23T23:40:13.7970915Z   File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.7971729Z   File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.7972400Z   File "h5py/h5f.pyx", line 88, in h5py.h5f.open
2021-06-23T23:40:13.7972983Z OSError: Unable to open file (file signature not found)
2021-06-23T23:40:13.7974306Z DEBUG    dandi:dandiapi.py:454 Calculated dandi-etag of c461208d22d3fd53da6f748284d1b49f-1 for /tmp/pytest-of-runner/pytest-0/text_dandiset23/subdir2/coconut.txt
2021-06-23T23:40:13.7975446Z DEBUG    dandi:dandiapi.py:463 subdir2/coconut.txt: Beginning upload
2021-06-23T23:40:13.7976278Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/uploads/initialize/
2021-06-23T23:40:13.7976989Z DEBUG    dandi:dandiapi.py:167 Response: 409
2021-06-23T23:40:13.7977884Z DEBUG    dandi:dandiapi.py:174 Error 409 while sending POST request to http://localhost:8000/api/uploads/initialize/: "Blob already exists."
2021-06-23T23:40:13.7978892Z DEBUG    dandi:dandiapi.py:478 subdir2/coconut.txt: Blob already exists on server
2021-06-23T23:40:13.7979814Z DEBUG    dandi:dandiapi.py:551 subdir2/coconut.txt: Assigning asset blob to dandiset & version
2021-06-23T23:40:13.7980744Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7981486Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7982257Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7983007Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7983696Z INFO     dandi:dandiapi.py:565 subdir2/coconut.txt: Asset successfully uploaded
2021-06-23T23:40:13.7984582Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.7985333Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.7986432Z ERROR    dandi:upload.py:249 Failed to extract metadata from /tmp/pytest-of-runner/pytest-0/text_dandiset23/subdir2/banana.txt
2021-06-23T23:40:13.7987252Z Traceback (most recent call last):
2021-06-23T23:40:13.7988278Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/upload.py", line 245, in process_path
2021-06-23T23:40:13.7989036Z     asset_metadata = nwb2asset(
2021-06-23T23:40:13.7990049Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 670, in nwb2asset
2021-06-23T23:40:13.7990836Z     metadata = get_metadata(nwb_path)
2021-06-23T23:40:13.7991877Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 116, in fingerprinter
2021-06-23T23:40:13.7992736Z     ret = fingerprinted(path, *args, **kwargs_)
2021-06-23T23:40:13.7993767Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 591, in __call__
2021-06-23T23:40:13.7994520Z     return self._cached_call(args, kwargs)[0]
2021-06-23T23:40:13.7995541Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 534, in _cached_call
2021-06-23T23:40:13.7996332Z     out, metadata = self.call(*args, **kwargs)
2021-06-23T23:40:13.7997329Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 761, in call
2021-06-23T23:40:13.7998080Z     output = self.func(*args, **kwargs)
2021-06-23T23:40:13.7999195Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 84, in fingerprinted
2021-06-23T23:40:13.7999992Z     return f(path, *args, **kwargs)
2021-06-23T23:40:13.8001156Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 51, in get_metadata
2021-06-23T23:40:13.8001947Z     meta["nwb_version"] = get_nwb_version(path)
2021-06-23T23:40:13.8003016Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/pynwb_utils.py", line 95, in get_nwb_version
2021-06-23T23:40:13.8003829Z     with h5py.File(filepath, "r") as h5file:
2021-06-23T23:40:13.8004826Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 406, in __init__
2021-06-23T23:40:13.8005602Z     fid = make_fid(name, mode, userblock_size,
2021-06-23T23:40:13.8006601Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 173, in make_fid
2021-06-23T23:40:13.8007366Z     fid = h5f.open(name, flags, fapl=fapl)
2021-06-23T23:40:13.8008046Z   File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.8008838Z   File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.8009525Z   File "h5py/h5f.pyx", line 88, in h5py.h5f.open
2021-06-23T23:40:13.8010110Z OSError: Unable to open file (file signature not found)
2021-06-23T23:40:13.8011432Z DEBUG    dandi:dandiapi.py:454 Calculated dandi-etag of 1f2406f328ccfc069d2844d6fa64e564-1 for /tmp/pytest-of-runner/pytest-0/text_dandiset23/subdir2/banana.txt
2021-06-23T23:40:13.8012567Z DEBUG    dandi:dandiapi.py:463 subdir2/banana.txt: Beginning upload
2021-06-23T23:40:13.8013459Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/uploads/initialize/
2021-06-23T23:40:13.8014188Z DEBUG    dandi:dandiapi.py:167 Response: 409
2021-06-23T23:40:13.8015069Z DEBUG    dandi:dandiapi.py:174 Error 409 while sending POST request to http://localhost:8000/api/uploads/initialize/: "Blob already exists."
2021-06-23T23:40:13.8016082Z DEBUG    dandi:dandiapi.py:478 subdir2/banana.txt: Blob already exists on server
2021-06-23T23:40:13.8016918Z DEBUG    dandi:dandiapi.py:551 subdir2/banana.txt: Assigning asset blob to dandiset & version
2021-06-23T23:40:13.8017821Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.8018572Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8019323Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.8020083Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8020773Z INFO     dandi:dandiapi.py:565 subdir2/banana.txt: Asset successfully uploaded
2021-06-23T23:40:13.8021652Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.8022404Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8023493Z ERROR    dandi:upload.py:249 Failed to extract metadata from /tmp/pytest-of-runner/pytest-0/text_dandiset23/subdir1/apple.txt
2021-06-23T23:40:13.8024328Z Traceback (most recent call last):
2021-06-23T23:40:13.8025339Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/upload.py", line 245, in process_path
2021-06-23T23:40:13.8026122Z     asset_metadata = nwb2asset(
2021-06-23T23:40:13.8027138Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 670, in nwb2asset
2021-06-23T23:40:13.8027904Z     metadata = get_metadata(nwb_path)
2021-06-23T23:40:13.8028960Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 116, in fingerprinter
2021-06-23T23:40:13.8029818Z     ret = fingerprinted(path, *args, **kwargs_)
2021-06-23T23:40:13.8030833Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 591, in __call__
2021-06-23T23:40:13.8031605Z     return self._cached_call(args, kwargs)[0]
2021-06-23T23:40:13.8032711Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 534, in _cached_call
2021-06-23T23:40:13.8033501Z     out, metadata = self.call(*args, **kwargs)
2021-06-23T23:40:13.8034513Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/joblib/memory.py", line 761, in call
2021-06-23T23:40:13.8035252Z     output = self.func(*args, **kwargs)
2021-06-23T23:40:13.8047179Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/fscacher/cache.py", line 84, in fingerprinted
2021-06-23T23:40:13.8048045Z     return f(path, *args, **kwargs)
2021-06-23T23:40:13.8049088Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/metadata.py", line 51, in get_metadata
2021-06-23T23:40:13.8049904Z     meta["nwb_version"] = get_nwb_version(path)
2021-06-23T23:40:13.8050968Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/dandi/pynwb_utils.py", line 95, in get_nwb_version
2021-06-23T23:40:13.8051774Z     with h5py.File(filepath, "r") as h5file:
2021-06-23T23:40:13.8052794Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 406, in __init__
2021-06-23T23:40:13.8053564Z     fid = make_fid(name, mode, userblock_size,
2021-06-23T23:40:13.8054584Z   File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/h5py/_hl/files.py", line 173, in make_fid
2021-06-23T23:40:13.8055346Z     fid = h5f.open(name, flags, fapl=fapl)
2021-06-23T23:40:13.8056011Z   File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.8056983Z   File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
2021-06-23T23:40:13.8057669Z   File "h5py/h5f.pyx", line 88, in h5py.h5f.open
2021-06-23T23:40:13.8058242Z OSError: Unable to open file (file signature not found)
2021-06-23T23:40:13.8059569Z DEBUG    dandi:dandiapi.py:454 Calculated dandi-etag of 61ad88b8426662e2466f70234dc02c4b-1 for /tmp/pytest-of-runner/pytest-0/text_dandiset23/subdir1/apple.txt
2021-06-23T23:40:13.8060662Z DEBUG    dandi:dandiapi.py:463 subdir1/apple.txt: Beginning upload
2021-06-23T23:40:13.8061483Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/uploads/initialize/
2021-06-23T23:40:13.8062217Z DEBUG    dandi:dandiapi.py:167 Response: 409
2021-06-23T23:40:13.8063104Z DEBUG    dandi:dandiapi.py:174 Error 409 while sending POST request to http://localhost:8000/api/uploads/initialize/: "Blob already exists."
2021-06-23T23:40:13.8064110Z DEBUG    dandi:dandiapi.py:478 subdir1/apple.txt: Blob already exists on server
2021-06-23T23:40:13.8064936Z DEBUG    dandi:dandiapi.py:551 subdir1/apple.txt: Assigning asset blob to dandiset & version
2021-06-23T23:40:13.8065858Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.8066607Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8067569Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.8068343Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8069019Z INFO     dandi:dandiapi.py:565 subdir1/apple.txt: Asset successfully uploaded
2021-06-23T23:40:13.8070009Z ------------------------------ Captured log call -------------------------------
2021-06-23T23:40:13.8070816Z DEBUG    dandi:dandiarchive.py:423 Parsing url http://localhost:8000/api/dandisets/000028
2021-06-23T23:40:13.8072422Z DEBUG    dandi:dandiarchive.py:536 Parsed into DandisetURL(api_url=AnyHttpUrl('http://localhost:8000/api', scheme='http', host='localhost', host_type='int_domain', port='8000', path='/api'), dandiset_id='000028', version_id=None)
2021-06-23T23:40:13.8073662Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/
2021-06-23T23:40:13.8074329Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8075078Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/info/
2021-06-23T23:40:13.8075901Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8076598Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/
2021-06-23T23:40:13.8077287Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8077991Z DEBUG    dandi:download.py:397 Updating dandiset.yaml from obtained dandiset metadata
2021-06-23T23:40:13.8079021Z DEBUG    dandi:dandiset.py:99 Found identifier DANDI:000028 in top level 'identifier'
2021-06-23T23:40:13.8079895Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/
2021-06-23T23:40:13.8080651Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8081943Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/993c6a66-34df-40a3-81d1-c516ee2f6831/
2021-06-23T23:40:13.8082843Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8084015Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/7c545b63-eb4c-4afe-9197-c54d36ea5e5d/
2021-06-23T23:40:13.8084932Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8086083Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/fd73f535-18fe-4ba4-b794-c320023a8cf6/
2021-06-23T23:40:13.8086979Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8088087Z DEBUG    dandi:dandiapi.py:141 GET http://localhost:8000/api/dandisets/000028/versions/draft/assets/d723f463-9a38-4d8a-9339-497b06be6c68/
2021-06-23T23:40:13.8088943Z DEBUG    dandi:dandiapi.py:167 Response: 200
2021-06-23T23:40:13.8089699Z DEBUG    dandi:download.py:667 Starting new download in new download directory
2021-06-23T23:40:13.8091067Z DEBUG    dandi:dandiapi.py:375 Starting download from http://localhost:8000/api/dandisets/000028/versions/draft/assets/993c6a66-34df-40a3-81d1-c516ee2f6831/download/
2021-06-23T23:40:13.8092457Z INFO     dandi:dandiapi.py:389 Asset 993c6a66-34df-40a3-81d1-c516ee2f6831 successfully downloaded
2021-06-23T23:40:13.8094038Z DEBUG    dandi:download.py:606 Verified that /tmp/pytest-of-runner/pytest-0/test_download_newest_version0/000028/file.txt has correct dandi-etag d5b215bbddf463221d1cc55c31098d40-1
2021-06-23T23:40:13.8095233Z DEBUG    dandi:download.py:667 Starting new download in new download directory
2021-06-23T23:40:13.8096640Z DEBUG    dandi:dandiapi.py:375 Starting download from http://localhost:8000/api/dandisets/000028/versions/draft/assets/7c545b63-eb4c-4afe-9197-c54d36ea5e5d/download/
2021-06-23T23:40:13.8098115Z INFO     dandi:dandiapi.py:389 Asset 7c545b63-eb4c-4afe-9197-c54d36ea5e5d successfully downloaded
2021-06-23T23:40:13.8099736Z DEBUG    dandi:download.py:606 Verified that /tmp/pytest-of-runner/pytest-0/test_download_newest_version0/000028/subdir2/coconut.txt has correct dandi-etag c461208d22d3fd53da6f748284d1b49f-1
2021-06-23T23:40:13.8100951Z DEBUG    dandi:download.py:667 Starting new download in new download directory
2021-06-23T23:40:13.8102341Z DEBUG    dandi:dandiapi.py:375 Starting download from http://localhost:8000/api/dandisets/000028/versions/draft/assets/fd73f535-18fe-4ba4-b794-c320023a8cf6/download/
2021-06-23T23:40:13.8103778Z INFO     dandi:dandiapi.py:389 Asset fd73f535-18fe-4ba4-b794-c320023a8cf6 successfully downloaded
2021-06-23T23:40:13.8105388Z DEBUG    dandi:download.py:606 Verified that /tmp/pytest-of-runner/pytest-0/test_download_newest_version0/000028/subdir2/banana.txt has correct dandi-etag 1f2406f328ccfc069d2844d6fa64e564-1
2021-06-23T23:40:13.8106584Z DEBUG    dandi:download.py:667 Starting new download in new download directory
2021-06-23T23:40:13.8107930Z DEBUG    dandi:dandiapi.py:375 Starting download from http://localhost:8000/api/dandisets/000028/versions/draft/assets/d723f463-9a38-4d8a-9339-497b06be6c68/download/
2021-06-23T23:40:13.8109294Z INFO     dandi:dandiapi.py:389 Asset d723f463-9a38-4d8a-9339-497b06be6c68 successfully downloaded
2021-06-23T23:40:13.8110898Z DEBUG    dandi:download.py:606 Verified that /tmp/pytest-of-runner/pytest-0/test_download_newest_version0/000028/subdir1/apple.txt has correct dandi-etag 61ad88b8426662e2466f70234dc02c4b-1
2021-06-23T23:40:13.8112196Z DEBUG    dandi:dandiapi.py:141 POST http://localhost:8000/api/dandisets/000028/versions/draft/publish/
2021-06-23T23:40:13.8112953Z DEBUG    dandi:dandiapi.py:167 Response: 400
2021-06-23T23:40:13.8113974Z ERROR    dandi:dandiapi.py:176 Error 400 while sending POST request to http://localhost:8000/api/dandisets/000028/versions/draft/publish/: "Dandiset metadata or asset metadata is not valid"
```
</details>
other broken runs could be found in PRs against staging, e.g. https://github.com/dandi/dandi-api/pull/353

--- a/dandi/dandiapi.py
+++ b/dandi/dandiapi.py
@@ -4,6 +4,7 @@ import os.path
 from pathlib import Path
 import re
 from threading import Lock
+from time import sleep
 from typing import Any, Callable, Dict, Iterator, Optional, Union, cast
 from urllib.parse import urlparse, urlunparse
 from xml.etree.ElementTree import fromstring
@@ -421,6 +422,16 @@ class RemoteDandiset(APIBase):
         `RemoteDandiset` with the `version` attribute set to the new published
         `Version`.
         """
+        lgr.debug("Waiting for Dandiset %s to complete validation ...", self.identifier)
+        while True:
+            r = self.client.get(f"{self.version_api_path}info/")
+            if r["status"] == "Valid":
+                break
+            elif r["status"] == "Invalid":
+                raise ValueError(
+                    f"Dandiset {self.identifier} is invalid: {r['validation_error']}"
+                )
+            sleep(0.5)
         return self.copy(
             update={
                 "version": Version.parse_obj(

--- a/dandi/dandiapi.py
+++ b/dandi/dandiapi.py
@@ -425,6 +425,10 @@ class RemoteDandiset(APIBase):
         lgr.debug("Waiting for Dandiset %s to complete validation ...", self.identifier)
         while True:
             r = self.client.get(f"{self.version_api_path}info/")
+            if "status" not in r:
+                # Running against older version of dandi-api that doesn't
+                # validate
+                break
             if r["status"] == "Valid":
                 break
             elif r["status"] == "Invalid":

--- a/dandi/metadata.py
+++ b/dandi/metadata.py
@@ -15,7 +15,7 @@ from .pynwb_utils import (
     ignore_benign_pynwb_warnings,
     metadata_cache,
 )
-from .utils import ensure_datetime, get_utcnow_datetime
+from .utils import ensure_datetime, get_mime_type, get_utcnow_datetime
 
 lgr = get_logger()
 
@@ -706,7 +706,7 @@ def get_default_metadata(path, digest=None, digest_type=None) -> models.BareAsse
         dateModified=dateModified,
         blobDateModified=blobDateModified,
         wasGeneratedBy=[get_generator(start_time, end_time)],
-        # encodingFormat # TODO
+        encodingFormat=get_mime_type(path),
     )
 
 

--- a/dandi/tests/data/dandiarchive-docker/docker-compose.yml
+++ b/dandi/tests/data/dandiarchive-docker/docker-compose.yml
@@ -42,7 +42,7 @@ services:
       DJANGO_MINIO_STORAGE_SECRET_KEY: minioSecretKey
       DJANGO_STORAGE_BUCKET_NAME: django-storage
       DJANGO_MINIO_STORAGE_MEDIA_URL: http://localhost:9000/django-storage
-      DJANGO_DANDI_SCHEMA_VERSION: 1.0.0-rc1
+      DJANGO_DANDI_SCHEMA_VERSION:
       DANDI_ALLOW_LOCALHOST_URLS: "1"
     ports:
       - "8000:8000"
@@ -77,7 +77,7 @@ services:
       DJANGO_MINIO_STORAGE_SECRET_KEY: minioSecretKey
       DJANGO_STORAGE_BUCKET_NAME: django-storage
       DJANGO_MINIO_STORAGE_MEDIA_URL: http://localhost:9000/django-storage
-      DJANGO_DANDI_SCHEMA_VERSION: 1.0.0-rc1
+      DJANGO_DANDI_SCHEMA_VERSION:
       DANDI_ALLOW_LOCALHOST_URLS: "1"
 
   minio:

--- a/dandi/tests/data/dandiarchive-docker/docker-compose.yml
+++ b/dandi/tests/data/dandiarchive-docker/docker-compose.yml
@@ -43,6 +43,7 @@ services:
       DJANGO_STORAGE_BUCKET_NAME: django-storage
       DJANGO_MINIO_STORAGE_MEDIA_URL: http://localhost:9000/django-storage
       DJANGO_DANDI_SCHEMA_VERSION: 1.0.0-rc1
+      DANDI_ALLOW_LOCALHOST_URLS: "1"
     ports:
       - "8000:8000"
 
@@ -77,6 +78,7 @@ services:
       DJANGO_STORAGE_BUCKET_NAME: django-storage
       DJANGO_MINIO_STORAGE_MEDIA_URL: http://localhost:9000/django-storage
       DJANGO_DANDI_SCHEMA_VERSION: 1.0.0-rc1
+      DANDI_ALLOW_LOCALHOST_URLS: "1"
 
   minio:
     image: minio/minio:latest

--- a/dandi/tests/fixtures.py
+++ b/dandi/tests/fixtures.py
@@ -9,6 +9,7 @@ import tempfile
 from time import sleep
 from uuid import uuid4
 
+from dandischema.consts import DANDI_SCHEMA_VERSION
 from dateutil.tz import tzutc
 import pynwb
 import pytest
@@ -181,6 +182,7 @@ def docker_compose_setup():
     env = dict(os.environ)
     env["DJANGO_DANDI_GIRDER_API_URL"] = "http://localhost:8080/api/v1"
     env["DJANGO_DANDI_GIRDER_API_KEY"] = "abc123"
+    env["DJANGO_DANDI_SCHEMA_VERSION"] = DANDI_SCHEMA_VERSION
 
     try:
         if create:

--- a/dandi/tests/fixtures.py
+++ b/dandi/tests/fixtures.py
@@ -271,7 +271,23 @@ def local_dandi_api(docker_compose_setup):
 
 @pytest.fixture()
 def text_dandiset(local_dandi_api, monkeypatch, tmp_path_factory):
-    d = local_dandi_api["client"].create_dandiset("Text Dandiset", {})
+    d = local_dandi_api["client"].create_dandiset(
+        "Text Dandiset",
+        {
+            "schemaKey": "Dandiset",
+            "name": "Text Dandiset",
+            "description": "A test text Dandiset",
+            "contributor": [
+                {
+                    "schemaKey": "Person",
+                    "name": "Wodder, John",
+                    "roleName": ["dcite:Author", "dcite:ContactPerson"],
+                }
+            ],
+            "license": ["spdx:CC0-1.0"],
+            "manifestLocation": ["https://github.com/dandi/dandi-cli"],
+        },
+    )
     dandiset_id = d.identifier
     dspath = tmp_path_factory.mktemp("text_dandiset")
     (dspath / dandiset_metadata_file).write_text(f"identifier: '{dandiset_id}'\n")

--- a/dandi/tests/test_dandiapi.py
+++ b/dandi/tests/test_dandiapi.py
@@ -29,7 +29,23 @@ def test_upload(local_dandi_api, simple1_nwb, tmp_path):
 
 def test_publish_and_manipulate(local_dandi_api, monkeypatch, tmp_path):
     client = local_dandi_api["client"]
-    d = client.create_dandiset("Test Dandiset", {})
+    d = client.create_dandiset(
+        "Test Dandiset",
+        {
+            "schemaKey": "Dandiset",
+            "name": "Text Dandiset",
+            "description": "A test text Dandiset",
+            "contributor": [
+                {
+                    "schemaKey": "Person",
+                    "name": "Wodder, John",
+                    "roleName": ["dcite:Author", "dcite:ContactPerson"],
+                }
+            ],
+            "license": ["spdx:CC0-1.0"],
+            "manifestLocation": ["https://github.com/dandi/dandi-cli"],
+        },
+    )
     dandiset_id = d.identifier
     upload_dir = tmp_path / "upload"
     upload_dir.mkdir()

--- a/dandi/tests/test_utils.py
+++ b/dandi/tests/test_utils.py
@@ -18,6 +18,7 @@ from ..utils import (
     flatten,
     flattened,
     get_instance,
+    get_mime_type,
     get_module_version,
     get_utcnow_datetime,
     is_same_time,
@@ -351,3 +352,28 @@ def test_get_module_version():
     assert get_module_version("dandi") == __version__
     assert get_module_version("pynwb") == pynwb.__version__
     assert get_module_version("abracadabra123") is None
+
+
+@pytest.mark.parametrize(
+    "filename,mtype",
+    [
+        ("foo.txt", "text/plain"),
+        ("foo", "application/octet-stream"),
+        ("foo.gz", "application/gzip"),
+        ("foo.tar.gz", "application/gzip"),
+        ("foo.tgz", "application/gzip"),
+        ("foo.taz", "application/gzip"),
+        ("foo.svg.gz", "application/gzip"),
+        ("foo.svgz", "application/gzip"),
+        ("foo.Z", "application/x-compress"),
+        ("foo.tar.Z", "application/x-compress"),
+        ("foo.bz2", "application/x-bzip2"),
+        ("foo.tar.bz2", "application/x-bzip2"),
+        ("foo.tbz2", "application/x-bzip2"),
+        ("foo.xz", "application/x-xz"),
+        ("foo.tar.xz", "application/x-xz"),
+        ("foo.txz", "application/x-xz"),
+    ],
+)
+def test_get_mime_type(filename, mtype):
+    assert get_mime_type(filename) == mtype

--- a/dandi/utils.py
+++ b/dandi/utils.py
@@ -10,6 +10,7 @@ import inspect
 import io
 import itertools
 import logging
+from mimetypes import guess_type
 import os
 import os.path as op
 from pathlib import Path
@@ -709,3 +710,22 @@ def abbrev_prompt(msg: str, *options: str) -> str:
         answer = input(msg).lower()
         if answer in options_map:
             return options_map[answer]
+
+
+def get_mime_type(filename: str, strict: bool = False) -> str:
+    """
+    Like `mimetypes.guess_type()`, except that if the file is compressed, the
+    MIME type for the compression is returned.  Also, the default return value
+    is now ``'application/octet-stream'`` instead of `None`.
+    """
+    mtype, encoding = guess_type(filename, strict)
+    if encoding is None:
+        return mtype or "application/octet-stream"
+    elif encoding == "gzip":
+        # application/gzip is defined by RFC 6713
+        return "application/gzip"
+        # There is also a "+gzip" MIME structured syntax suffix defined by RFC
+        # 8460; exactly when can that be used?
+        # return mtype + '+gzip'
+    else:
+        return "application/x-" + encoding

